Propagate typed exceptions from replicas to coordinator

piodul opened the pull request:
This PR adds infrastructure which allows replicas to report more detailed information about errors to the coordinators. This ability will be useful for implementing per-partition rate limiting, as replicas will report to the coordinator that the rate limit has been reached and coordinator will propagate this information to the user.

A set of "replica-side exceptions" is introduced. Before propagating an exception to the coordinator, the replica converts it into one of the replica-side exceptions, then serializes and sends through RPC.

Current RPC related to regular reads and writes is modified to use the new replica-side exceptions.

Tests: unit(dev)

piodul commented:
Currently marking as draft because I wanted to push something today, but it is still not finished: a small number of todos is left, and the read-rpc related commit does not take into account the TYPED_ERRORS_IN_READ_RPC cluster feature. I will fix those tomorrow. Nevertheless, the changes won't be big so I invite everyone interested to review.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4978,7 +4965,7 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
                     }
                     slogger.log(l, "Failed to apply mutation from {}#{}: {}", reply_to, shard, eptr);
-                    errors++;
+                    errors.local = replica::encode_replica_exception(std::move(eptr));
```
Note updating errors.count?

avikivity commented on idl/storage_proxy.idl.hh:
```diff
@@ -11,9 +11,9 @@ verb [[with_client_info, one_way]] mutation_done (unsigned shard, uint64_t respo
-verb [[with_client_info, with_timeout]] read_data (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]];
-verb [[with_client_info, with_timeout]] read_mutation_data (query::read_command cmd, ::compat::wrapping_partition_range pr) -> reconcilable_result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]];
-verb [[with_client_info, with_timeout]] read_digest (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result_digest, api::timestamp_type [[version 1.2.0]], cache_temperature [[version 2.0.0]];
+verb [[with_client_info, with_timeout]] read_data (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 4.7.0]];
```
Hmm. optional<> adds another layer of meta-stuff. On the other hand it allows a quick exit on no error.

avikivity commented:
I like the series, but I guess no benefit until the replica and coordinator are converted to not throw in the first place.

piodul commented:
> I like the series, but I guess no benefit until the replica and coordinator are converted to not throw in the first place.

The main benefit here is that now we, at least partially, preserve information about exception types when reporting them to coordinator. It is a preparation for per-parition rate limiting in case of which the coordinator will have to react differently to replicas reporting per-partition limit being reached.

I agree that - by itself - it doesn't offer much feature- and performance-wise, but wanted to send it for review as it forms a logical whole. Should I resend it with more commits so that it gives more benefit, e.g. after resultizing the coordinator or just with the rest of the per-partition limiting?

piodul commented:
v2:
- read RPC handler now sends exception in new encoding but only if TYPED_ERRORS_IN_READ_RPC feature is enabled
- write coordinator now properly chooses if the exception should be printed or not, with the same algorithm as before
- refactored the `encode_replica_exception` function a bit in order to implement the previous point
- changed 4.7.0 to 5.1.0 in `storage_proxy.idl.hh`
- fixed a small bug where the local replica error wasn't accounted in the error count

avikivity commented:
> > I like the series, but I guess no benefit until the replica and coordinator are converted to not throw in the first place.
> 
> The main benefit here is that now we, at least partially, preserve information about exception types when reporting them to coordinator. It is a preparation for per-parition rate limiting in case of which the coordinator will have to react differently to replicas reporting per-partition limit being reached.

Ok. A worthwhile goal.

> 
> I agree that - by itself - it doesn't offer much feature- and performance-wise, but wanted to send it for review as it forms a logical whole. Should I resend it with more commits so that it gives more benefit, e.g. after resultizing the coordinator or just with the rest of the per-partition limiting?

If it's simple enough then I'd like to see the reader_concurrency_semaphore timeout path converted. This is a massive win that helps non-partition-rate-limiter users.

The entry point appears to be

```c++
void reader_concurrency_semaphore::expiry_handler::operator()(entry& e) noexcept {
    e.pr.set_exception(named_semaphore_timed_out(_semaphore._name));

    maybe_dump_reader_permit_diagnostics(_semaphore, _semaphore._permit_list, "timed out");
}
```

and the non-timeout path is forward_to() in reader_concurrency_semaphore::execution_loop(). So it might not be too difficult. /cc @denesb

avikivity commented:
(if it turns out too complicated we can defer it to a next step)

piodul commented:
I tried to track how the timeouts are propagated from the semaphore and I'm afraid the amount of work to resultize all of the paths might be non-trivial.

While there are two uses of `with_permit` (in `database::query` and `database::query_mutations`) which should be straightforward to resultize, I can also see that a permit is obtained in the `shard_reader_v2`. Getting rid of exceptions there would probably require the whole flat mutation reader interface to be adjusted, which sounds like a lot of work.

I'd rather postpone it to a later step.

piodul commented:
(Moving out of draft, as I fixed all the issues that I initially wanted to fix before de-drafting)

avikivity commented on idl/storage_proxy.idl.hh:
```diff
@@ -11,9 +11,9 @@ verb [[with_client_info, one_way]] mutation_done (unsigned shard, uint64_t respo
-verb [[with_client_info, with_timeout]] read_digest (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result_digest, api::timestamp_type [[version 1.2.0]], cache_temperature [[version 2.0.0]];
+verb [[with_client_info, with_timeout]] read_data (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 5.1.0]];
+verb [[with_client_info, with_timeout]] read_mutation_data (query::read_command cmd, ::compat::wrapping_partition_range pr) -> reconcilable_result [[lw_shared_ptr]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 5.1.0]];
+verb [[with_client_info, with_timeout]] read_digest (query::read_command cmd, ::compat::wrapping_partition_range pr, query::digest_algorithm digest [[version 3.0.0]]) -> query::result_digest, api::timestamp_type [[version 1.2.0]], cache_temperature [[version 2.0.0]], std::optional<replica::exception_variant> [[version 5.1.0]];
```
Strictly speaking we should have returned a `variant<reconcilable_result, exception_variant>, cache_temperature`, but that requires a new verb. Let's keep it as is, but when we are forced to add a new verb, let's remember to change it then.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5123,12 +5132,22 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
+                if (!f.failed()) {
+                    auto&& [result, temp] = f.get();
+                    return make_ready_future<final_tuple_type>(std::move(result), std::move(temp), std::nullopt);
+                } else if (features().cluster_supports_typed_errors_in_read_rpc()) {
```
So far we usually check features on the coordinator side, not replica. The reasoning is that features aren't enabled simultaneously on all nodes. I don't think it's a problem here.

avikivity replied:
(we usually have the coordinator enable a bit in the request, and the replica checks the request-specific bit)

@psarna can you think of a reason why checking on the replica would be problematic?

psarna replied:
Our features are a one-way ticket, once the whole cluster supports it, there's no way back. So it sounds perfectly fine to check the feature on a replica, because if it returns true, its coordinator should also be aware of it.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -5154,12 +5173,22 @@ void storage_proxy::init_messaging_service(shared_ptr<migration_manager> mm) {
+                if (!f.failed()) {
+                    auto&& [result, temp] = f.get();
+                    return make_ready_future<final_tuple_type>(std::move(result), std::move(temp), std::nullopt);
+                } else if (features().cluster_supports_typed_errors_in_read_rpc()) {
```
Why do you want to convert existing exceptions to the new interface? You previous series allowed to propagate some errors as exceptions and others as values. I think here we can do the same. Rate limiting errors can be propagated as values, but all others stay as they are. Also for rate limiting would it be useful to allow propagating rate limiting requests along with a successful reply?

avikivity replied:
Converting exceptions to values protects against exception storms reducing throughput.

gleb-cloudius replied:
Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).

Another point is that requesting a rate limit can be done together with returning a valid result.

avikivity replied:
This is for timeout exceptions. Not RPC timeouts, but timeouts detected on the replica while waiting. You're right that we need to change RPC to return std::expected (or something similar) too.

I don't understand the second point.

piodul replied:
> Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).

This PR is only concerned with encoding exceptions and transferring them without losing information about the type. I see "rate limit exceeded" as a kind of an error response, similar to "timeout" or others. I thought it would make sense to introduce a common mechanism for reporting them.

I'm aware that I am converting back to exception, I intend to get rid of this problem in later PRs. This one is just about preserving the types.

> Another point is that requesting a rate limit can be done together with returning a valid result.

Maybe, although currently I plan for the "rate limit exceeded" signal to always mean an error without a result. If we add a possibility to request rate limit while returning the result, we will have to signal the second case differently from the first anyway.

gleb-cloudius replied:
> This is for timeout exceptions. Not RPC timeouts, but timeouts detected on the replica while waiting. You're right that we need to change RPC to return std::expected (or something similar) too.

Timeouts detected on a replica side are never reach a coordinator. They are dropped by the client when sending a reply since rpc sees that the execution timed out, so for the timeout converting it to something else here is just a waste of cpu time.

> 
> I don't understand the second point.

Rate limit request does not have to be an error. Piotr seams to understood what I mean.

gleb-cloudius replied:
> > Yes, but with previous patches we decided that we do it for exceptions that usually cause storms like timeout exception (hence using `std::expected` or similar. Here for instance we do not solve the problem of timeout exception at all since it is reported on a coordinator by the rpc as an exception anyway. I understand that rate limiting can cause storms and it worth to be propagated as a value (BTW it look like it is converted back to an exception right after been reserved).
> 
> This PR is only concerned with encoding exceptions and transferring them without losing information about the type. I see "rate limit exceeded" as a kind of an error response, similar to "timeout" or others. I thought it would make sense to introduce a common mechanism for reporting them.
> 

And timeout exceptions will never be sent back, so converting them just a waste of time. Some others are rare (like request aborted).

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/92/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/benipeled_test/10/

piodul commented:
I'm closing because I was convinced by Gleb that timeouts/unspecified failures do not need to be converted and without it I no longer think that this PR makes sense by itself. I'll send the modified functionality along with the rest of per-partition limiting.
